modules/monitoring/files/usr/lib/nagios/plugins/check_cdn_log_s3_freshness

```diff
@@ -0,0 +1,122 @@
+strings for example :(
+
+The S3 object names look like, for example:
+
+govuk_assets/year=2020/month=01/date=30/2020-01-30T22:00:00.000-DXdcYBSfUdJyVvwbPcAn.log.gz
+
+For help on command-line options:
+    check_cdn_log_s3_freshness -h
+
+Example usage for testing the script:
+    check_cdn_log_s3_freshness -e staging -t 2020-01-31T17:00
+
+Example usage in production:
+    check_cdn_log_s3_freshness -e production -l govuk_www -c 1
+"""
+
+import argparse
+import datetime
+import enum
+import logging
+import sys
+
+import boto3
+
+
+class IcingaStatus(enum.IntEnum):
+    """Exit statuses for Icinga checks."""
+    OK = 0
+    WARNING = 1
+    CRITICAL = 2
+
+
+def fromisoformat(s):
+    """Parse a string YYYY-MM-DDTHH:MM and return a datetime. Assumes UTC.
+
+    This is only meant for passing in a fake time from the command-line when
+    testing. Would use datetime.fromisoformat, but we're stuck on Python 3.4.
+    """
+    return datetime.datetime.strptime(s, '%Y-%m-%dT%H:%M')
+
+
+def parse_args():
+    """Return an argparse.Namespace populated with command-line args."""
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-e', '--env', default='production',
+                        help='Environment to check: integration, staging, production.')
+    parser.add_argument('-l', '--log_type', default='govuk_assets',
+                        help='Which logs to check: govuk_assets, govuk_www.')
+    parser.add_argument('-c', '--critical_age_hours', type=int, default=1,
+                        help='If the newest logs are older than this many hours, '
+                             'return CRITICAL status.')
+    parser.add_argument('-t', '--time', type=fromisoformat,
+                        help='For testing purposes, use the given time as if it\'s the current '
+                             'time. Requires the format YYYY-MM-DDTHH:MM. Assumes UTC.')
+    parser.add_argument('-v', '--verbose', action='count',
+                        help='Show DEBUG log messages.')
+    return parser.parse_args()
+
+
+def get_s3_bucket(env):
+    """Given the environment name, return the boto S3 resource for the bucket."""
+    s3 = boto3.resource('s3')
+    bucket_name = 'govuk-%s-fastly-logs' % env
+    logging.info('S3 bucket name: %s', bucket_name)
+    return s3.Bucket(bucket_name)
+
+
+def get_prefix(log_type, critical_age_hours, fake_time=None):
+    """Return the S3 prefix to be searched for log files, based on the time."""
+    max_age = datetime.timedelta(hours=critical_age_hours)
+    now = fake_time or datetime.datetime.now(datetime.timezone.utc)  # TZ-aware datetime in UTC.
+    query_time = now - max_age  # Oldest acceptable last_modified time for logs.
+    time_prefix = query_time.strftime('year=%Y/month=%m/date=%d/%Y-%m-%dT%H')
+    logging.info('Time is %s%s', query_time.isoformat(),
+                 ' (overridden for testing)' if fake_time else '')
+    return '%s/%s' % (log_type, time_prefix)
```

Comment:
I'm wondering if we might have an integration and staging issue where it looks up data that doesn't exist. I'm not sure I understand exactly how this works so I'll explain my understanding in case that is flawed.

So I understand that every 60s the icinga check runs. Assuming the time is 19:03 today and we are in production we would send a query to S3 to check whether there are any files in the bucket with the following prefix: "year=2020/month=02/date=14/2020-02-14T18" where the 18 is because it is current time subtracted by one hour.

So assuming I'm correct there (which is a big assumption 😄) if we assume we're in integration and there is no logs between 2:00am and 3:00am. If the time is 8:15am on February 14th then would we not be querying for a prefix subtracted by 6 hours? which would be  "year=2020/month=02/date=14/2020-02-14T02" which is a time where we have 0 results?

Reply:
I think you're absolutely right there. My initial approach was too simplistic and indeed it was just plain broken in the Integration case 🤦‍♂  Thanks for spotting it and for the nice clear explanation!

I've reworked it to parse the time of the latest log file based on the filename and compare that against the threshold. Sorry it's taken me ages to get around to it. As a bonus, we've now got better info logging and 1-minute precision instead of 1 hour.

Ready for another look, if you have a spare moment. Thanks!